Add support to MPS

andimarafioti opened the pull request:
I worked on adding support to MPS. During this work, I also did a bit of refactoring. 

Namely the changes are:

- Adding more modularity by adding new types of models
- Adding LightningWhisperMLX since it works considerably better on MPS
- Adding MeloTTS to have a more lightweight solution for TTS, more adequate for macbooks
- Add a "Local audio streamer", to avoid messing with connections to a localhost, which proved difficult on my macbook.
- Bypass bugs that I found while testing with mac (mainly having to generate all the text from the LLM and then synchronize the memory before generating with the TTS)

andimarafioti commented:
I also added support to pass mps as a device, see https://github.com/huggingface/speech-to-speech/issues/18

andimarafioti commented:
Also mentioned here: https://github.com/huggingface/speech-to-speech/issues/9

RonanKMcGovern commented:
Just tested and having the same issues as with [here](https://github.com/huggingface/speech-to-speech/issues/18)

Are there any additional installations required?

What command do you recommend for running? I'm using:
```
git clone https://github.com/eustlb/speech-to-speech.git
cd speech-to-speech
pip install -r requirements.txt
pip install git+https://github.com/nltk/nltk.git@3.8.2
python s2s_pipeline.py --recv_host localhost --send_host localhost --lm_model_name HuggingFaceTB/SmolLM-360M-Instruct --lm_device mps --stt_device mps --tts_device mps
```

Many thanks.

andimarafioti commented:
First, run it with:

`python s2s_pipeline.py --mode local --device mps`

Second, you need to install MeloTTS. For that you can add this to your script. 

```bash
git clone https://github.com/myshell-ai/MeloTTS.git
cd MeloTTS
pip install -e .
python -m unidic download
```

Let me know if that works!

andimarafioti commented:
Looking again at your code, to test this code you need to checkout the branch as well! 
`git checkout add-mps-support`

RonanKMcGovern commented:
Many thanks @andimarafioti ,making good progress now with this:
```
git clone https://github.com/eustlb/speech-to-speech.git
cd speech-to-speech
git fetch --all
git checkout add-mps-support
python -m venv mpsEnv 
source mpsEnv/bin/activate
pip install --upgrade pip
pip install -r requirements.txt
pip install --upgrade transformers
pip install git+https://github.com/nltk/nltk.git@3.8.2
git clone https://github.com/myshell-ai/MeloTTS.git
cd MeloTTS
pip install -e .
python -m unidic download
cd ../
python s2s_pipeline.py --mode local --device mps --lm_model_name HuggingFaceTB/SmolLM-360M-Instruct
```

takes quite a while at this point:
```
python s2s_pipeline.py --mode local --device mps                               
/Users/ronanmcgovern/TR/mpsEnv/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
2024-08-19 17:48:47,969 - parler_tts.modeling_parler_tts - WARNING - Flash attention 2 is not installed
Using cache found in /Users/ronanmcgovern/.cache/torch/hub/snakers4_silero-vad_master
python(10073) MallocStackLogging: can't turn off malloc stack logging because it was not enabled.
Fetching 13 files: 100%|███████████████| 13/13 [00:00<00:00, 35614.60it/s]
```

What are the default models btw, and what are your recommendations for reducing RAM requirements.

I see distil-whisper is there in large v3 form, which is 1.5 GB - not too bad. I reckon whisper small is good enough (I use it for youtube transcriptions). And I see you are defaulting to smollm and parler jenny so those are small already.


Nice work, thanks

mitkox commented:
What is the recommended python version?

DFin commented:
I tried this but so far wasnt successful: 

 ```
> git clone https://github.com/eustlb/speech-to-speech.git
> cd speech-to-speech
> git fetch --all
> git checkout add-mps-support
> python -m venv mpsEnv 
> source mpsEnv/bin/activate
> pip install --upgrade pip
> pip install -r requirements.txt
> pip install --upgrade transformers
> pip install git+https://github.com/nltk/nltk.git@3.8.2
> git clone https://github.com/myshell-ai/MeloTTS.git
> cd MeloTTS
> pip install -e .
> python -m unidic download
> cd ../
> python s2s_pipeline.py --mode local --device mps --lm_model_name HuggingFaceTB/SmolLM-360M-Instruct
> 
```

Initially i missed both a Rust compiler and mecap. I installed mecab: 

```
brew install mecab                                      
brew install mecab-ipadic
```

and Rust compiler using:
```
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
. "$HOME/.cargo/env"     
```

But tiktoken still wont compile. Rust compiler is giving out an error about wrong reference to mutable variable (see full ouput below): 
```
error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
```

I guess I could run the compiler with allowing `invalid_reference_casting` but TBH don't know enough about Rust to feel comfortable doing that. Are there any precompiled packages for tiktoken or does someone else know a better solution. 


Full error message:

``` 
      warning: variable does not need to be mutable
         --> tokenizers-lib/src/models/unigram/model.rs:265:21
          |
      265 |                 let mut target_node = &mut best_path_ends_at[key_pos];
          |                     ----^^^^^^^^^^^
          |                     |
          |                     help: remove this `mut`
          |
          = note: `#[warn(unused_mut)]` on by default
      
      warning: variable does not need to be mutable
         --> tokenizers-lib/src/models/unigram/model.rs:282:21
          |
      282 |                 let mut target_node = &mut best_path_ends_at[starts_at + mblen];
          |                     ----^^^^^^^^^^^
          |                     |
          |                     help: remove this `mut`
      
      warning: variable does not need to be mutable
         --> tokenizers-lib/src/pre_tokenizers/byte_level.rs:200:59
          |
      200 |     encoding.process_tokens_with_offsets_mut(|(i, (token, mut offsets))| {
          |                                                           ----^^^^^^^
          |                                                           |
          |                                                           help: remove this `mut`
      
      error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
         --> tokenizers-lib/src/models/bpe/trainer.rs:526:47
          |
      522 |                     let w = &words[*i] as *const _ as *mut _;
          |                             -------------------------------- casting happend here
      ...
      526 |                         let word: &mut Word = &mut (*w);
          |                                               ^^^^^^^^^
          |
          = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
          = note: `#[deny(invalid_reference_casting)]` on by default
      
      warning: `tokenizers` (lib) generated 3 warnings
      error: could not compile `tokenizers` (lib) due to 1 previous error; 3 warnings emitted
```

RonanKMcGovern commented:
Yeah, there have been updates it seems since yesterday and issues now with
tokenizers. I'm seeing the same.

On Tue, Aug 20, 2024 at 10:59 AM David Finsterwalder <
***@***.***> wrote:

> I tried this but so far wasnt successful:
>
> > git clone https://github.com/eustlb/speech-to-speech.git
> > cd speech-to-speech
> > git fetch --all
> > git checkout add-mps-support
> > python -m venv mpsEnv
> > source mpsEnv/bin/activate
> > pip install --upgrade pip
> > pip install -r requirements.txt
> > pip install --upgrade transformers
> > pip install ***@***.***
> > git clone https://github.com/myshell-ai/MeloTTS.git
> > cd MeloTTS
> > pip install -e .
> > python -m unidic download
> > cd ../
> > python s2s_pipeline.py --mode local --device mps --lm_model_name HuggingFaceTB/SmolLM-360M-Instruct
> >
>
> Initially i missed both a Rust compiler and mecap. I installed mecab:
>
> brew install mecab
> brew install mecab-ipadic
>
> and Rust compiler using:
>
> curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
> . "$HOME/.cargo/env"
>
> But tiktoken still wont compile. Rust compiler is giving out an error
> about wrong reference to mutable variable (see full ouput below):
>
> error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
>
> I guess I could run the compiler with allowing invalid_reference_casting
> but TBH don't know enough about Rust to feel comfortable doing that. Are
> there any precompiled packages for tiktoken or does someone else know a
> better solution.
>
> Full error message:
>
>       warning: variable does not need to be mutable
>          --> tokenizers-lib/src/models/unigram/model.rs:265:21
>           |
>       265 |                 let mut target_node = &mut best_path_ends_at[key_pos];
>           |                     ----^^^^^^^^^^^
>           |                     |
>           |                     help: remove this `mut`
>           |
>           = note: `#[warn(unused_mut)]` on by default
>
>       warning: variable does not need to be mutable
>          --> tokenizers-lib/src/models/unigram/model.rs:282:21
>           |
>       282 |                 let mut target_node = &mut best_path_ends_at[starts_at + mblen];
>           |                     ----^^^^^^^^^^^
>           |                     |
>           |                     help: remove this `mut`
>
>       warning: variable does not need to be mutable
>          --> tokenizers-lib/src/pre_tokenizers/byte_level.rs:200:59
>           |
>       200 |     encoding.process_tokens_with_offsets_mut(|(i, (token, mut offsets))| {
>           |                                                           ----^^^^^^^
>           |                                                           |
>           |                                                           help: remove this `mut`
>
>       error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
>          --> tokenizers-lib/src/models/bpe/trainer.rs:526:47
>           |
>       522 |                     let w = &words[*i] as *const _ as *mut _;
>           |                             -------------------------------- casting happend here
>       ...
>       526 |                         let word: &mut Word = &mut (*w);
>           |                                               ^^^^^^^^^
>           |
>           = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
>           = note: `#[deny(invalid_reference_casting)]` on by default
>
>       warning: `tokenizers` (lib) generated 3 warnings
>       error: could not compile `tokenizers` (lib) due to 1 previous error; 3 warnings emitted
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/speech-to-speech/pull/20#issuecomment-2298465491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASVG6CXO26X4DIEVR5OFFC3ZSMHR7AVCNFSM6AAAAABMXZZL26VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOJYGQ3DKNBZGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

andimarafioti commented:
Let me check this quickly

andimarafioti commented:
I tried the bash script you uploaded and it also doesn't work for me but with another error (pip can't install MuCab, no clue why). I'm using uv (https://github.com/astral-sh/uv/) mainly because it's super fast, but in this case it does solve the dependency better. Another thing I noticed in your bash script that's different from mine is that I reinstall the latest transformers and tokenizers at the end, because melotts downgrades them. There is a more elegant way to do this and I'll work on it to close this PR, but I'm currently finishing the tech report for Idefics3 😅 The final thing that I broke with my last commit is that I  made parlerTTS the default tts and it uses a global variable that's declared on the whisper model. That's pretty hacky, I'll fix it as well.
 
The bash script that worked for me is:

```bash
git clone https://github.com/eustlb/speech-to-speech.git
cd speech-to-speech
git fetch --all
git checkout add-mps-support
uv venv --python 3.11
source .venv/bin/activate
uv pip install -r requirements.txt
cd MeloTTS
uv pip install -e .
python -m unidic download
cd ..
uv pip install -r requirements.txt  # melotts downgrades transformers...
python s2s_pipeline.py --mode local --device mps --lm_model_name HuggingFaceTB/SmolLM-360M-Instruct  --tts melo
```

Please let me know if this works for you as well!

andimarafioti commented:
I just noticed this might still break for you if you are missing some nltk files on your machine. That one is hard to test, but if you run this in docker it will probably happen.

BIGdeadLock commented:
I'm also working on it on #26 

My current error after after some changes (some were mentioned here) is:
```
ImportError: cannot import name 'TextIteratorStreamer' from 'transformers' (/Users/eden.yavin/Projects/speech-to-speech/venv/lib/python3.9/site-packages/transformers/__init__.py)

```

andimarafioti commented:
Yes, you need t update transformers again 😅  because melotts downgrades it. You can re-do uv pip install -r requirements.txt after installing melotts

scalar27 commented:
I tried to follow the examples in this thread and have gotten s2s to work.  I'm getting a lot of slow downs and stuttering when I use Parler.  I'm on an M1 Max 64Gb.  Is this to be expected or should it perform better than this?

BIGdeadLock commented:
@scalar27 

Try using Melo : `--tts melo`. It made the audio much better when I switched to it

scalar27 commented:
> @scalar27
> 
> Try using Melo : `--tts melo`. It made the audio much better when I switched to it

Yes, that is much faster but I specifically would like to try to use Parler if it's reasonably responsive.

andimarafioti commented:
Hi! To make Parler work, you can increase the "play_steps_s" parameter we pass to it. Then, your chunks are larger, so naturally, the audio is less chunky. With my setup (M3 Pro, 36 GB), I couldn't get Parler to be above real-time, so it always chops up since it needs time for the next chunk to be ready after playing the last one. I'm trying to convince the Parler team to train a smaller model for this precise use case. Fingers crossed!

RonanKMcGovern commented:
Here are install commands that work for me:
```
```

When I run those, the TTS seems to default to parler. I changed the default from parler to melo in s2s_pipeline.py and then get:
```
python s2s_pipeline.py --mode local --device mps --lm_model_name HuggingFaceTB/SmolLM-360M-Instruct  --tts melo
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/ronanmcgovern/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
Using cache found in /Users/ronanmcgovern/.cache/torch/hub/snakers4_silero-vad_master
Fetching 13 files: 100%|█████████████████████| 13/13 [00:00<00:00, 25491.33it/s]
```
at which point I make no further progress.

1. @andimarafioti thanks for your work on this. When you've a chance, could you ensure that passing --tts melo overrides the default?
2. Any ideas on why things are freezing above? Do you know the total VRAM required? I'm guessing it's about 1.5 for whisper + 0.700 for Smol + 0.400 for Melo. So maybe 3.6 GB? Or more if any of this is in 32 bits... probably does get challenging for a Mac M1 with 8 GB RAM.

BIGdeadLock commented:
@RonanKMcGovern 

When I passed the --melo TTS I noticed a big difference in the speech performance so I assume it works and override the default.

The thing I did find is that passing the -lm_mode_name to SMOLM-360M did not override the Phi3 default which is drastically bigger.

andimarafioti commented:
I pushed a change to let you use the SMOLLM, I had MLX-LM by default which is optimized for mac, but doesn't implement the SMOLLM. For the melo tts, it should work if you pass it :S The hangup you are experiencing sounds to me like it's taking too long to load, it might be because of some off loading of models on initialization. I can recommend you to install asitop to monitor what's going on under the hood: https://github.com/tlkh/asitop

andimarafioti commented:
I fixed the install 🥳 Now it should work with `uv pip install -r requirements.txt`. That will install Melo from my fork at: https://github.com/andimarafioti/MeloTTS
There is after the step to do: `python -m unidic download` if you want to use Melo, I didn't add that to the install because it's a lot of overhead for people that don't want melo, but I added a warning if you get the error you should get without it. I also fixed a bunch of things that were open. I'm going to close this PR with this to focus on other improvements I want to do to the codebase, please do share if any issues arise and I'll address them promptly.

scalar27 commented:
> Hi! To make Parler work, you can increase the "play_steps_s" parameter we pass to it. Then, your chunks are larger, so naturally, the audio is less chunky. With my setup (M3 Pro, 36 GB), I couldn't get Parler to be above real-time, so it always chops up since it needs time for the next chunk to be ready after playing the last one. I'm trying to convince the Parler team to train a smaller model for this precise use case. Fingers crossed!

Are we already using Parler mini?  https://huggingface.co/parler-tts/parler-tts-mini-v1

andimarafioti commented:
Yes, the default is "ylacombe/parler-tts-mini-jenny-30H"

BIGdeadLock commented:
@andimarafioti 

The installation with uv seems to work well.

But, after typing the command to run it I'm getting an error with melo

```
Traceback (most recent call last):
  File "/Users/eden.yavin/Projects/speech-to-speech/s2s_pipeline.py", line 1143, in <module>
    main()
  File "/Users/eden.yavin/Projects/speech-to-speech/s2s_pipeline.py", line 1119, in main
    from TTS.melotts import MeloTTSHandler
  File "/Users/eden.yavin/Projects/speech-to-speech/TTS/melotts.py", line 1, in <module>
    from melo.api import TTS
  File "/Users/eden.yavin/Projects/speech-to-speech/venv/lib/python3.9/site-packages/melo/api.py", line 18, in <module>
    from .download_utils import load_or_download_config, load_or_download_model
  File "/Users/eden.yavin/Projects/speech-to-speech/venv/lib/python3.9/site-packages/melo/download_utils.py", line 4, in <module>
    from cached_path import cached_path
  File "/Users/eden.yavin/Projects/speech-to-speech/venv/lib/python3.9/site-packages/cached_path/__init__.py", line 13, in <module>
    from ._cached_path import cached_path
  File "/Users/eden.yavin/Projects/speech-to-speech/venv/lib/python3.9/site-packages/cached_path/_cached_path.py", line 15, in <module>
    from .schemes import (
  File "/Users/eden.yavin/Projects/speech-to-speech/venv/lib/python3.9/site-packages/cached_path/schemes/__init__.py", line 6, in <module>
    from .r2 import R2Client
  File "/Users/eden.yavin/Projects/speech-to-speech/venv/lib/python3.9/site-packages/cached_path/schemes/r2.py", line 18, in <module>
    class R2Client(SchemeClient):
  File "/Users/eden.yavin/Projects/speech-to-speech/venv/lib/python3.9/site-packages/cached_path/schemes/r2.py", line 20, in R2Client
    botocore.exceptions.HTTPClientError,
AttributeError: module 'botocore.exceptions' has no attribute 'HTTPClientError'

```

storybe reviewed (commented):
Test
